spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/ComputeTableStatsSparkAction.java

```diff
@@ -0,0 +1,150 @@
+
+  private final Table table;
+  private Set<String> columns;
+  private Snapshot snapshot;
+
+  ComputeTableStatsSparkAction(SparkSession spark, Table table) {
+    super(spark);
+    this.table = table;
+    this.snapshot = table.currentSnapshot();
+    this.columns =
+        table.schema().columns().stream().map(Types.NestedField::name).collect(Collectors.toSet());
+  }
+
+  @Override
+  protected ComputeTableStatsSparkAction self() {
+    return this;
+  }
+
+  @Override
+  public Result execute() {
+    String desc = String.format("Computing stats for %s", table.name());
+    JobGroupInfo info = newJobGroupInfo("COMPUTE-TABLE-STATS", desc);
+    return withJobGroupInfo(info, this::doExecute);
+  }
+
+  private Result doExecute() {
+    if (snapshot == null) {
+      return ImmutableComputeTableStats.Result.builder().build();
+    }
+    LOG.info("Computing stats of {} for snapshot {}", table.name(), snapshot.snapshotId());
+    List<Blob> blobs = generateNDVBlobs();
+    StatisticsFile statisticFile;
+    try {
+      statisticFile = writeAndCommitPuffin(blobs);
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+    return ImmutableComputeTableStats.Result.builder().statisticsFile(statisticFile).build();
+  }
+
+  private StatisticsFile writeAndCommitPuffin(List<Blob> blobs) throws IOException {
+    LOG.info(
+        "Writing stats to puffin files for table {} for snapshot{}",
+        table.name(),
+        snapshot.snapshotId());
+    TableOperations operations = ((HasTableOperations) table).operations();
+    String path = operations.metadataFileLocation(String.format("%s.stats", UUID.randomUUID()));
+    OutputFile outputFile = operations.io().newOutputFile(path);
+    GenericStatisticsFile statisticsFile = generateStatisticsFile(blobs, outputFile, path);
+    table.updateStatistics().setStatistics(snapshot.snapshotId(), statisticsFile).commit();
+    return statisticsFile;
+  }
+
+  @NotNull
+  private GenericStatisticsFile generateStatisticsFile(
+      List<Blob> blobs, OutputFile outputFile, String path) throws IOException {
+    try (PuffinWriter writer =
+        Puffin.write(outputFile).createdBy("Iceberg ComputeTableStats action").build()) {
+      blobs.forEach(writer::add);
+      writer.finish();
+      return new GenericStatisticsFile(
+          snapshot.snapshotId(),
+          path,
+          writer.fileSize(),
+          writer.footerSize(),
+          writer.writtenBlobsMetadata().stream()
+              .map(GenericBlobMetadata::from)
+              .collect(ImmutableList.toImmutableList()));
+    }
+  }
+
+  private List<Blob> generateNDVBlobs() {
+    return NDVSketchGenerator.generateNDVSketchesAndBlobs(spark(), table, snapshot, columns);
+  }
+
+  @Override
+  public ComputeTableStats columns(String... columnNames) {
+    Preconditions.checkArgument(
+        columnNames != null && columnNames.length > 0, "Columns cannot be null/empty");
+    for (String columnName : columnNames) {
```

Comment:
I don't think we can validate columns at this point as the call that sets snapshot ID may come later. It is only safe to validate the columns are valid during the execution (e.g. in `doExecute`).

I guess it can be something like this:

```
@Override
public ComputeTableStats columns(String... newColumns) {
  Preconditions.checkArgument(
      newColumns != null && newColumns.length > 0,
      "Columns to compute stats can't be null/empty");
  this.columns = ImmutableSet.copyOf(newColumns);
  return this;
}

private void validateColumns() {
  Schema schema = table.schemas().get(snapshot.schemaId());
  for (String column : columns()) {
    Types.NestedField field = schema.findField(column);
    Preconditions.checkArgument(field != null, "Column not found: %s", column);
  }
}
```

And then call `validateColumns()` in `doExecute()`.